### ISSUE TITLE
Function passed as parameter and saved in Context API store takes old store from Context API

### ISSUE THREAD

**Issue by Ghost180195** [facebook/react#17723]
When I get 401 status code from backend I run refreshToken method with passing the function where expired token occurred. In refreshToken method I get new token and set in refreshTokenLastFunc property function from parameter.

Then I watch when refreshTokenLastFunc was updated using React useEffect and run once again the function where expired token occurred.

The problem is while I run store.refreshTokenLastFunc() in useEffect, the function in refreshTokenLastFunc property uses old Context API store(so it uses old token not the new one). You can read my comment in useEffect for store.refreshTokenLastFunc.

**Do you want to request a *feature* or report a *bug*?**
I don't actually know if it is a React bug or I don't understand JS properly

**What is the current behavior?**
If I ran **store.refreshTokenLastFunc**  function in useEffect it takes old store, not the new one

**What is the expected behavior?**
I want **store.refreshTokenLastFunc**  function to run with new store, because store was modified before this function run

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
React verison: 16.12.0
Browser: Google Chrome: 79

My code:
```js
export const StoreProvider = props => {
  const getToken = () => localStorage.getItem("token");

  const initState = () => ({
    token: getToken(),
    isAuth: false,
    userRole: "old role",
    mainUrl: MainUrl,
    apiUrl: ApiUrl,
    refreshTokenLastFunc: () => {}
  });

  const [store, setStore] = useState(initState());

  const getUserInfo = async () => {
    if (getToken()) {
      try {
        const apiConfig = {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${store.token}`
          }
        };

        const response = await fetch(`${store.apiUrl}get-me`, apiConfig);
        const responseJson = await response.json();

        if (response.ok) {
          // Update Context API
          setStore({
            ...store,
            userRole: responseJson.role,
            userName: responseJson.name,
            userGroupId: responseJson.group_id,
            isAuth: true
          });
        } else if (response.status === 401) {
          refreshToken(getUserInfo);
        } else {
          throw new Error(`Some error occurred`);
        }
      } catch (error) {
        console.log(error);
      }
    }
  };

  const refreshToken = async func => {
    try {
      const apiConfig = {
        method: "GET",
        headers: {
          Accept: "application/json",
          Authorization: `Bearer ${store.token}`
        }
      };

      const response = await fetch(`${store.mainUrl}refresh-token`, apiConfig);
      const responseJson = await response.json();

      if (response.ok) {
        // Update token in local storage
        localStorage.setItem("token", responseJson.token);
        // Update Context API
        setStore({
          ...store,
          userRole: "new role",
          token: responseJson.token,
          refreshTokenLastFunc: func
        });
      } else {
        throw new Error(`Some error...`);
      }
    } catch (error) {
      throw error;
    }
  };

  useEffect(() => {
    getUserInfo();
  }, []);

  useEffect(() => {
    // If I console log my store before calling function, store is correctly updated, but the function uses old store.
    console.log("store from useEffect: ", store); // store.userRole = 'new role' which is correct
    store.refreshTokenLastFunc(); // store.userRole = 'old role' which should be 'new role'
  }, [store.refreshTokenLastFunc]);

  return (
    <StoreContext.Provider value={[store, setStore, logout, getUserInfo]}>
      {props.children}
    </StoreContext.Provider>
  );
};
```

**Comment by kunukn** [comment]
Can you use the functional update instead?
https://reactjs.org/docs/hooks-reference.html#functional-updates

e.g.

```jsx
setStore(prevStore => ({
    ...prevStore,
    userRole: 'new role',
    token: responseJson.token,
    refreshTokenLastFunc: func,
  })
)
```

**Comment by bvaughn** [comment]
This looks like it boils down to an issue of [how "scope" works in JavaScript](https://scotch.io/tutorials/understanding-scope-in-javascript).

Drilling in on this comment in particular:
```js
// If I console log my store before calling function, store is correctly updated, but the function uses old store.
console.log("store from useEffect: ", store); // store.userRole = 'new role' which is correct
store.refreshTokenLastFunc(); // store.userRole = 'old role' which should be 'new role'
```

The reason you see different values being logged is that you are logging two different `store` values.
1. The `store` you are logging in `useEffect` is the most recently updated state value. It's the one you *want* to be logging.
1. The `store` you are logging inside of `refreshTokenLastFunc` is a *previous* `store` value, because the `refreshTokenLastFunc` function (really just `getUserInfo`) closed over an *earlier* version of `store`.

I've added some comments to `refreshToken` to explain what's happening:
```js
setStore({
  // This is the version of "store" that `refreshTokenLastFunc` sees
  ...store,
  // Your new role goes into a new version of `store`,
  // along with `refreshTokenLastFunc`
  userRole: "new role",
  token: responseJson.token,
  refreshTokenLastFunc: func
});
```

I don't really understand why your code is structured this way, but maybe the easiest solution would be to move `getUserInfo` *outside of your component* and just pass `store` and `setStore` as parameters to it.

P.S. I've edited your original question to fix the markdown formatting so that it's easier for others to read.
